### PR TITLE
[SPARK-15638][SQL] Audit Dataset, SparkSession, and SQLContext

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This patch contains a list of changes as a result of my auditing Dataset, SparkSession, and SQLContext. The patch audits the categorization of experimental APIs, function groups, and deprecations. For the detailed list of changes, please see the diff.
## How was this patch tested?

N/A

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala [hunk 1]
```diff
@@ -52,7 +52,7 @@ case class CatalogStorageFormat(
 
 object CatalogStorageFormat {
   /** Empty storage format for default values and copies. */
-  val EmptyStorageFormat = CatalogStorageFormat(locationUri = None, inputFormat = None,
+  val empty = CatalogStorageFormat(locationUri = None, inputFormat = None,
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala]: cc @andrewor14 @viirya 

from 

https://github.com/apache/spark/pull/13327#discussion_r64943992

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileCatalogSuite.scala [hunk 2]
```diff
@@ -19,6 +19,8 @@ package org.apache.spark.sql.execution.datasources
 
 import java.io.File
 
+import scala.language.reflectiveCalls
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileCatalogSuite.scala]: removes a bunch of compile time warnings

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDDLCommandSuite.scala [hunk 3]
```diff
@@ -37,9 +35,9 @@ class HiveDDLCommandSuite extends PlanTest {
 
   private def extractTableDesc(sql: String): (CatalogTable, Boolean) = {
     parser.parsePlan(sql).collect {
-      case CreateTableCommand(desc, allowExisting) => (desc, allowExisting)
-      case CreateTableAsSelectLogicalPlan(desc, _, allowExisting) => (desc, allowExisting)
-      case CreateViewCommand(desc, _, allowExisting, _, _, _) => (desc, allowExisting)
+      case c: CreateTableCommand => (c.table, c.ifNotExists)
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveDDLCommandSuite.scala]: related to https://github.com/apache/spark/pull/13327#discussion_r64944224

File: sql/core/src/test/scala/org/apache/spark/sql/SQLContextSuite.scala [hunk 4]
```diff
@@ -24,6 +24,7 @@ import org.apache.spark.sql.catalyst.TableIdentifier
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types.{BooleanType, StringType, StructField, StructType}
 
+@deprecated("This suite is deprecated to silent compiler deprecation warnings", "2.0.0")
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLContextSuite.scala]: removes a bunch of compile time warnings

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 5]
```diff
@@ -41,15 +41,13 @@ import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Project}
...
  *                 unless they are specified with full qualified table name with database prefix.
- * @param sql the original sql
  */
 case class CreateViewCommand(
     tableDesc: CatalogTable,
     child: LogicalPlan,
     allowExisting: Boolean,
     replace: Boolean,
-    isTemporary: Boolean,
-    sql: String)
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: related to https://github.com/apache/spark/pull/13327#discussion_r64944224

cc @andrewor14 @viirya

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 6]
```diff
@@ -132,6 +130,15 @@ class Column(protected[sql] val expr: Expression) extends Logging {
     case _ => UnresolvedAttribute.quotedString(name)
   })
 
+  override def toString: String = usePrettyExpression(expr).sql
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Column => that.expr.equals(this.expr)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: how about `that.expr.semanticEquals(this.expr)`? One column equals to another if they always produce same result.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: oh sorry you just move them up, it's fine to keep them same as before.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 7]
```diff
@@ -204,8 +205,8 @@ class KeyValueGroupedDataset[K, V] private[sql](
    * Internal helper function for building typed aggregations that return tuples.  For simplicity
    * and code reuse, we do this without the help of the type system and then use helper functions
    * that cast appropriately for the user facing interface.
-   * TODO: does not handle aggregations that return nonflat results,
    */
+  // TODO: does not handle aggregations that return nonflat results.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: I'm pretty sure this TODO is already done.

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 7 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: cool i will remove it

**Comment by rxin** [comment]
Thanks - merging in master/2.0.

**Pull request closed without merging**